### PR TITLE
Compile Scala library with Dotty and test its TASTy

### PR DESCRIPTION
**Pull request by nicolasstucki**

### CODE REVIEW

File: dotty-scala-library-from-tasty-tests/test/LibraryTest.scala [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+    "scala.StringContext",
+    "scala.sys.process.ProcessImpl",
+
+    // undefined: _root_.scala.StringContext.apply(["Scala "," "," -- ","" : String]:String*).f # -1: TermRef(TypeRef(TermRef(ThisType(TypeRef(NoPrefix,module class <root>)),module scala),class StringContext),f) at readTasty
+    "scala.collection.immutable.Range",
+    "scala.collection.Iterator",
+    "scala.util.PropertiesTrait",
+  )
+
+  def compileBlacklist = List[String](
```
Comment (smarter) [inline dotty-scala-library-from-tasty-tests/test/LibraryTest.scala:206]: We already compile the stdlib as part of the community build, so how come we're seeing compile failures here?

Comment (nicolasstucki) [reply dotty-scala-library-from-tasty-tests/test/LibraryTest.scala:126]: I added some docs

Comment (nicolasstucki) [reply dotty-scala-library-from-tasty-tests/test/LibraryTest.scala:133]: Those recompilations from the tasty of the stdlib

File: project/Build.scala [hunk 2]
```diff
@@ -817,6 +817,74 @@ object Build {
       javaOptions := (javaOptions in `dotty-compiler-bootstrapped`).value
     )
 
+  lazy val `dotty-scala-library-from-tasty-tests` = project.in(file("dotty-scala-library-from-tasty-tests")).
```
Comment (smarter) [inline project/Build.scala:820]: I don't understand what this project is for.

File: project/Build.scala [hunk 3]
```diff
@@ -817,6 +817,74 @@ object Build {
...
+    withCommonSettings(Bootstrapped).
+    dependsOn(`dotty-tasty-inspector` % "test->test").
+    settings(commonBootstrappedSettings).
+    settings(
+      javaOptions := (javaOptions in `dotty-compiler-bootstrapped`).value,
+      javaOptions += "-Ddotty.scala.library=" + packageBin.in(`dotty-scala-library`, Compile).value.getAbsolutePath
+    )
+
+
+  // Scala library compiled by dotty using the latest published sources of the library
```
Comment (smarter) [inline project/Build.scala:830]: I suggest using the stdlib sources from our community-build instead, that way if we need to change something we can change it in the community build submodule.

File: project/Build.scala [hunk 4]
```diff
@@ -817,6 +817,74 @@ object Build {
...
+        file.getPath.endsWith("scala/Any.scala") ||
+        file.getPath.endsWith("scala/AnyVal.scala") ||
+        file.getPath.endsWith("scala/AnyRef.scala") ||
+        file.getPath.endsWith("scala/Nothing.scala") ||
+        file.getPath.endsWith("scala/Null.scala") ||
+        file.getPath.endsWith("scala/Singleton.scala")))
+    )
+
+  // Tests that we can compile and run an app using the scala-library compiled by dotty
+  lazy val `dotty-scala-library-sandbox` = project.in(file("sandbox/dotty-scala-library")).
```
Comment (smarter) [inline project/Build.scala:881]: can we use `dotty-scala-library/test` instead of adding an extra project?

File: project/Build.scala [hunk 5]
```diff
@@ -817,6 +817,74 @@ object Build {
...
+    withCommonSettings(Bootstrapped).
+    dependsOn(`dotty-scala-library`).
+    settings(
+      managedClasspath in Compile ~= {
+        _.filterNot(file => file.data.getName == "scala-library-2.13.3.jar")
```
Comment (smarter) [inline project/Build.scala:886]: If you want to remove the stdlib from the dependencies of a scala project, you can use `autoScalaLibrary := false` (this will also remove dotty-library)

File: dotty-scala-library/src/scala/StringContext.scala [hunk 6]
```diff
@@ -0,0 +1,488 @@
+/*
```
Comment (smarter) [inline dotty-scala-library/src/scala/StringContext.scala:1]: Why is this file needed? I tthought we already handled `macro ???` to allow compiling the scala 2 stdlb as is.

**Review by smarter — CHANGES_REQUESTED** [review]

File: project/Build.scala (hunk 3 above)
Comment (nicolasstucki) [reply project/Build.scala:830]: Those sources are not aligned with the library that we are using to run dotty apps. Their documentation may also slightly defer.

File: project/Build.scala (hunk 4 above)
Comment (nicolasstucki) [reply project/Build.scala:881]: The library dependency setup is different on those two projects. That would not be testing what we want to test.

File: project/Build.scala (hunk 5 above)
Comment (nicolasstucki) [reply project/Build.scala:886]: We don't want to remove the dotty-library.

File: project/Build.scala (hunk 3 above)
Comment (smarter) [reply project/Build.scala:830]: We should keep them aligned if possible.

File: project/Build.scala (hunk 4 above)
Comment (smarter) [reply project/Build.scala:881]: Different how?

File: dotty-scala-library/src/scala/StringContext.scala (hunk 6 above)
Comment (nicolasstucki) [reply dotty-scala-library/src/scala/StringContext.scala:1]: We can compile them, but we assume that if they are going to be consumed from Scala 3 we need a scala 3 macro implementation as well. Here that second implementation is missing and when we load the class we ignore the Scala 2 macros. This implies that when loaded, StringContext does not have `s`, `f` and `raw` methods.

The usual solution is to add the second macro, but because those are intrinsic macros we can get away by just making them non-macros and let the compiler do its usual transformation of these methods.

File: project/Build.scala (hunk 5 above)
Comment (smarter) [reply project/Build.scala:886]: Ok, but then don't hardcore 2.13.3, use `stdlibVersion`

File: dotty-scala-library/src/scala/StringContext.scala (hunk 6 above)
Comment (smarter) [reply dotty-scala-library/src/scala/StringContext.scala:1]: Can't we interpret `macro ???` like `???` when it doesn't come from scala 2 so that we can keep using the same source?

File: project/Build.scala (hunk 2 above)
Comment (nicolasstucki) [reply project/Build.scala:820]: It is there to test that we can recompile the jar from `dotty-scala-library` from a bootstrapped compiler with the normal bootstrapped library. We needed an extra project to get the dependencies on the bootstrapped compiler right.

File: dotty-scala-library/src/scala/StringContext.scala (hunk 6 above)
Comment (nicolasstucki) [reply dotty-scala-library/src/scala/StringContext.scala:1]: Not under the normal dual macros rules. We would have to add an extra compilation flag for this special case which seems to be overkill in this single case.

File: project/Build.scala (hunk 4 above)
Comment (nicolasstucki) [reply project/Build.scala:881]: I don't remember why we did not set up the tests of that project to this. @sjrd was there a specific reason? Or just for simplicity?

File: project/Build.scala (hunk 3 above)
Comment (sjrd) [reply project/Build.scala:830]: If you change things because you need to change something, by definition they're not aligned anymore.

Comment (smarter) [reply project/Build.scala:830]: Sure but we can make PRs to keep them aligned if needed. I think it's just confusing to operate on multiple versions of scala-library at the same time, I don't see the advantages.

File: dotty-scala-library/src/scala/StringContext.scala (hunk 6 above)
Comment (smarter) [reply dotty-scala-library/src/scala/StringContext.scala:1]: > which seems to be overkill in this single case.

I disagree, I think that a special case is better than having a copy of file we need to worry about keeping up to date. Furthermore, it would be nice if in scala/scala we didn't just compile the stdlib with dotty but we also ran the stdlib tests, that is only possible if the compiler can handle the normal StringContext correctly if I understand this correctly.

File: project/Build.scala (hunk 3 above)
Comment (nicolasstucki) [reply project/Build.scala:830]: Under that logic, if we don't want to operate on multiple versions of the scala-library, we should remove the stdlib213 from the community-build as it is not aligned with the one we use to run dotty. But the one in the community build allows us to future proof the compilation of the library, therefore it is good to have this second variant of the scala-library.

File: dotty-scala-library/src/scala/StringContext.scala (hunk 6 above)
Comment (nicolasstucki) [reply dotty-scala-library/src/scala/StringContext.scala]: I removed the file and special-cased in these macros in `TreeUnpickler`.

**Comment by nicolasstucki** [comment]
@smarter I merged the sandbox into `scala3-scala2-library` (former `dotty-scala-library`). I kept `scala3-scala2-library-tasty-tasty-tests` (former `dotty-scala-library-tasty-tests`). 
I tried to put the tests in `BootstrappedOnlyCompilationTests` but they were timing out all the time on the CI, these tests are a bit too heavy to run in parallel with other tests.

**Comment by smarter** [comment]
`scala3-scala2-library` isn't a great name, maybe `stdlib-bootstrapped`?

File: project/Build.scala [hunk 7]
```diff
@@ -805,6 +805,65 @@ object Build {
...
+        file.getPath.endsWith("scala-library-src/scala/AnyRef.scala") ||
+        file.getPath.endsWith("scala-library-src/scala/Nothing.scala") ||
+        file.getPath.endsWith("scala-library-src/scala/Null.scala") ||
+        file.getPath.endsWith("scala-library-src/scala/Singleton.scala"))),
+      managedClasspath in Test ~= {
+        _.filterNot(file => file.data.getName == s"scala-library-${stdlibVersion(Bootstrapped)}.jar")
+      },
+    )
+
+  lazy val `scala3-scala2-library-tasty-tests` = project.in(file("scala3-scala2-library-tasty-tests")).
```
Comment (smarter) [inline project/Build.scala:858]: This should have some documentation explaining what it tests and why that needs to be done in a separate project.

Comment (nicolasstucki) [reply project/Build.scala:867]: Added

**Comment by nicolasstucki** [comment]
@smarter now it is ready for re-review. All tests pass again.

File: project/Build.scala [hunk 8]
```diff
@@ -805,6 +805,74 @@ object Build {
       javaOptions := (javaOptions in `scala3-compiler-bootstrapped`).value
     )
 
+  /** Scala library compiled by dotty using the latest published sources of the library */
+  lazy val `scala3-scala2-library` = project.in(file("scala3-scala2-library")).
```
Comment (smarter) [inline project/Build.scala:809]: I'm not a fan of that name, I suggest `stdlib-booststrapped` or `stdlib-recompiled` maybe.

File: scala3-scala2-library-tasty-tests/test/Scala3Scala2LibraryTASYyTest.scala [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+  /** Test that we can load trees from TASTy */
+  @Test def testTastyInspector: Unit =
+    loadWithTastyInspector(loadBlacklisted)
+
+  /** Test that we can load and compile trees from TASTy */
+  @Test def testFromTasty: Unit =
+    compileFromTasty(loadBlacklisted.union(compileBlacklisted))
+
+  @Ignore
+  @Test def testWhiteListFromTasty: Unit =
```
Comment (smarter) [inline scala3-scala2-library-tasty-tests/test/Scala3Scala2LibraryTASYyTest.scala:27]: What do we need a whitelist for?

File: scala3-scala2-library-tasty-tests/test/Scala3Scala2LibraryTASYyTest.scala [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+      assert(list.size == set.size,
+        list.diff(set.toSeq).mkString(s"`$name` has duplicate entries:\n  ", "\n  ", "\n\n"))
+    testDup("loadBlacklist", loadBlacklist, loadBlacklisted)
+    testDup("compileBlacklist", compileBlacklist, compileBlacklisted)
+
+  @Test def blacklistsNoIntersection =
+    val intersection = loadBlacklisted & compileBlacklisted
+    assert(intersection.isEmpty,
+      intersection.mkString(
+        "`compileBlacklist` contains names that are already in `compileBlacklist`: \n  ", "\n  ", "\n\n"))
```
Comment (smarter) [inline scala3-scala2-library-tasty-tests/test/Scala3Scala2LibraryTASYyTest.scala:44]: That error message mentions "compileBlacklist" twice when it probably should mention "loadBlacklist"?

File: scala3-scala2-library-tasty-tests/test/Scala3Scala2LibraryTASYyTest.scala [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+
+  /** List of classes that cannot be loaded from TASTy */
+  def loadBlacklist = List[String](
+    // No issues :)
+  )
+
+  /** List of classes that cannot be recompilied from TASTy */
+  def compileBlacklist = List[String](
+    // OK?
+    // failed: java.lang.AssertionError: assertion failed: class Boolean
```
Comment (smarter) [inline scala3-scala2-library-tasty-tests/test/Scala3Scala2LibraryTASYyTest.scala:135]: We should have an issue for thos and the comment should point to that issue

**Review by smarter — CHANGES_REQUESTED** [review]

File: scala3-scala2-library-tasty-tests/test/Scala3Scala2LibraryTASYyTest.scala (hunk 9 above)
Comment (nicolasstucki) [reply scala3-scala2-library-tasty-tests/test/Scala3Scala2LibraryTASYyTest.scala:27]: It was a simple way to test a subset of files. Now that I have minimized everything I can remove it.

File: scala3-scala2-library-tasty-tests/test/Scala3Scala2LibraryTASYyTest.scala (hunk 10 above)
Comment (nicolasstucki) [reply scala3-scala2-library-tasty-tests/test/Scala3Scala2LibraryTASYyTest.scala:44]: Now that the blacklist has shrunk I we can do without checking it.

Comment (nicolasstucki) [reply scala3-scala2-library-tasty-tests/test/Scala3Scala2LibraryTASYyTest.scala:44]: indeed

File: scala3-scala2-library-tasty-tests/test/Scala3Scala2LibraryTASYyTest.scala (hunk 9 above)
Comment (nicolasstucki) [reply scala3-scala2-library-tasty-tests/test/Scala3Scala2LibraryTASYyTest.scala:27]: On second thought, I prefer to keep in case a change breaks the compilation of the library. This would help to minimize the failing combination of files faster.

Comment (smarter) [reply scala3-scala2-library-tasty-tests/test/Scala3Scala2LibraryTASYyTest.scala:27]: That's fine but then please add a comment on top of the test explaining that

**Comment by nicolasstucki** [comment]
`scala3` prefix is not needed but it makes naming consistent

**Comment by smarter** [comment]
I think ideally we would only use the `scala3` prefix for the artifacts we distribute (library, compiler, ...), for internal things it doesn't seem useful.

**Comment by nicolasstucki** [comment]
We already have a few projects with `scala3` that we don’t distribute: `scala3-tastydoc-input`, `scala3-bench-run`, `scala3-sbt-bridge-tests`, `scala3-tastydoc`, `scala3-library`, `scala3-compiler`, `scala3-bench`
Should we rename those?

**Comment by smarter** [comment]
>  Should we rename those?

We could yes (in fact, we could also rename scala3-compiler to compiler, etc, we just need the artifact name to start with scala3-, but the name of the sbt project can be whatever we want, so we could make them easier to type)

**Comment by nicolasstucki** [comment]
Let's make a single PR afterward to renames them. It will be simpler to track.

**Comment by nicolasstucki** [comment]
I renamed this one to `stdlib-bootstrapped`.

File: project/Build.scala [hunk 12]
```diff
@@ -805,6 +805,74 @@ object Build {
       javaOptions := (javaOptions in `scala3-compiler-bootstrapped`).value
     )
 
+  /** Scala library compiled by dotty using the latest published sources of the library */
+  lazy val `stdlib-bootstrappedject.in(file("stdlib-stdlib-bootstrapped
```
Comment (smarter) [inline project/Build.scala:809]: typo

File: project/Build.scala [hunk 13]
```diff
@@ -805,6 +805,74 @@ object Build {
...
+  /** Test the tasty generated by `stdlib-bootstrapped
+   *
+   *  The tests are run with the bootstrapped compiler and the tasty inpector on the classpath.
+   *  The classpath has the default `scala-library` and not `stdlib-bootstrapped
+   *
+   *  The jar of `stdlib-bootstrappedovided for to the tests.
+   *   - inspector: test that we can load the contents of the jar using the tasty inspector
+   *   - from-tasty: test that we can recompile the contents of the jar using `dotc -from-tasty`
+   */
+  lazy val `stdlib-bootstrappedtests` = project.in(file("stdlib-stdlib-bootstrapped).
```
Comment (smarter) [inline project/Build.scala:867]: I wouldn't call that project just stdlib-bootstrappedtests, because it's not clear how it's different from stdlib-bootstrapped/test, -tasty-test or -sanity-test would be better I think

File: project/Build.scala [hunk 14]
```diff
@@ -805,6 +805,74 @@ object Build {
       javaOptions := (javaOptions in `scala3-compiler-bootstrapped`).value
     )
 
+  /** Scala library compiled by dotty using the latest published sources of the library */
+  lazy val `stdlib-bootstrappedject.in(file("stdlib-stdlib-bootstrapped
+    withCommonSettings(Bootstrapped).
+    dependsOn(dottyCompiler(Bootstrapped) % "provided; compile->runtime; test->test").
+    dependsOn(`scala3-tasty-inspector` % "test->test").
+    settings(commonBootstrappedSettings).
+    settings(
```
Comment (smarter) [inline project/Build.scala:814]: You could add a setting:
```suggestion
    settings(
      moduleName := "scala-library"
```
this way the jar will be called scala-library_3.0.0-M1.jar which makes sense.

Comment (nicolasstucki) [reply project/Build.scala:814]: Added

File: project/Build.scala (hunk 13 above)
Comment (nicolasstucki) [reply project/Build.scala]: fixed

File: project/Build.scala (hunk 12 above)
Comment (nicolasstucki) [reply project/Build.scala]: fixed

**Review by smarter — APPROVED** [review]
